### PR TITLE
Validate duplicated pathset entries in the permission files and fix permission ordering

### LINKED ISSUE #85: Investigate why the latest release changes the order of permissions in the higher privileged permissions column.

### PR DESCRIPTION
**Pull request by wandeg**
Duplicate pathset entries were resulting in some docs being rendered incorrectly. Fixes #85 and #87
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/microsoftgraph/kibali/pull/89)

### CODE REVIEW

File: test/kibaliTests/LeastPrivilegeTests.cs [hunk 1]
```diff
@@ -233,7 +233,6 @@ private PermissionsDocument CreatePermissionsDocument()
                             },
                             Paths = {
                                 { "/foo",  "least=DelegatedWork,Application" },
-                                { "/bar",  null },
                                 { "/bar/{id}",  null },
```
Comment (wandeg) [inline test/kibaliTests/LeastPrivilegeTests.cs:236]: Removing duplicate pathset entry

File: src/kibali/AcceptableClaimComparer.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+            if (x.AlsoRequires == null || y.AlsoRequires == null)
+                return false;
+
+            // Compare the contents of the AlsoRequires arrays
+            return x.AlsoRequires.SequenceEqual(y.AlsoRequires);
+        }
+
+        public int GetHashCode(AcceptableClaim obj)
+        {
+            if (obj == null || obj.Permission == null)
```
Comment (SilasKenneth) [inline src/kibali/AcceptableClaimComparer.cs:33]:
```suggestion
            if (obj?.Permission == null)
```

File: src/kibali/AcceptableClaimComparer.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+                return 0;
+
+            int hash = obj.Permission.GetHashCode();
+
+            if (obj.AlsoRequires != null)
+            {
+                foreach (var item in obj.AlsoRequires)
+                {
+                    hash = hash ^ item.GetHashCode();
```
Comment (SilasKenneth) [inline src/kibali/AcceptableClaimComparer.cs:42]:
```suggestion
                    hash ^= item.GetHashCode();
```

File: src/kibali/AcceptableClaimComparer.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
Comment (SilasKenneth) [inline src/kibali/AcceptableClaimComparer.cs:5]:
```suggestion
```

File: src/kibali/AcceptableClaimComparer.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
+﻿using System;
```
Comment (SilasKenneth) [inline src/kibali/AcceptableClaimComparer.cs:1]:
```suggestion
```

File: src/kibali/ProtectedResource.cs [hunk 6]
```diff
@@ -55,7 +55,15 @@ public void AddRequiredClaims(string permission, PathSet pathSet, string[] least
                         claim.SupportedEnvironments = provisioningInfo.Environment?.Split(";").ToList();
                         claim.IsEnabled = provisioningInfo.IsEnabled;
                     }
-                    acceptableClaims.Add(claim);
+                    if (!acceptableClaims.Add(claim))
+                    {
+                        errors.Add(new PermissionsError
+                        {
+                            Path = this.Url,
```
Comment (SilasKenneth) [inline src/kibali/ProtectedResource.cs:62]:
```suggestion
                            Path = Url,
```

File: src/kibali/ProtectedResource.cs [hunk 7]
```diff
@@ -55,7 +55,15 @@ public void AddRequiredClaims(string permission, PathSet pathSet, string[] least
...
+                            ErrorCode = PermissionsErrorCode.DuplicatePathsetEntry,
+                            Message = string.Format(StringConstants.DuplicatePathsetEntryErrorMessage, permission, this.Url, schemeKey, supportedMethod),
```
Comment (SilasKenneth) [inline src/kibali/ProtectedResource.cs:64]:
```suggestion
                            Message = string.Format(StringConstants.DuplicatePathsetEntryErrorMessage, permission, Url, schemeKey, supportedMethod),
```

File: src/kibali/ProtectedResource.cs [hunk 8]
```diff
@@ -113,20 +122,33 @@ public HashSet<PermissionsError> ValidateMismatchedSchemes(string permission, Pa
...
-                    existingScheme.AddRange(newScheme.Value);
+                    foreach (var entry in newScheme.Value)
+                    {
+                        if (!existingScheme.Add(entry))
+                        {
+                            duplicateErrors.Add(new PermissionsError
+                            {
+                                Path = this.Url,
+                                ErrorCode = PermissionsErrorCode.DuplicatePathsetEntry,
+                                Message = string.Format(StringConstants.DuplicatePathsetEntryErrorMessage, permission, this.Url, newScheme.Key, supportedMethod),
```
Comment (SilasKenneth) [inline src/kibali/ProtectedResource.cs:140]:
```suggestion
                                Message = string.Format(StringConstants.DuplicatePathsetEntryErrorMessage, permission, Url, newScheme.Key, supportedMethod),
```

File: src/kibali/ProtectedResource.cs [hunk 9]
```diff
@@ -113,20 +122,33 @@ public HashSet<PermissionsError> ValidateMismatchedSchemes(string permission, Pa
...
                 if (existingSchemes.TryGetValue(newScheme.Key, out var existingScheme))
                 {
-                    existingScheme.AddRange(newScheme.Value);
+                    foreach (var entry in newScheme.Value)
+                    {
+                        if (!existingScheme.Add(entry))
+                        {
+                            duplicateErrors.Add(new PermissionsError
+                            {
+                                Path = this.Url,
```
Comment (SilasKenneth) [inline src/kibali/ProtectedResource.cs:138]:
```suggestion
                                Path = Url,
```

File: src/kibali/ProtectedResource.cs [hunk 10]
```diff
@@ -406,7 +428,7 @@ private static void ProcessPermissionPairs(HashSet<(string, string)> permissionP
             }
             if (higherPrivilegedPairs.Count != 0)
             {
-                higher = string.Join(", ", higherPrivilegedPairs.OrderByDescending(x => x.Length));
+                higher = string.Join(", ", higherPrivilegedPairs.OrderBy(x => x));
```
Comment (millicentachieng) [inline src/kibali/ProtectedResource.cs:431]: @wandeg, Is the ordering here necessary? We still have a huge diff with this change. https://github.com/microsoftgraph/microsoft-graph-docs/pull/24946/files

Comment (wandeg) [reply src/kibali/ProtectedResource.cs]: Updated

**Review by millicentachieng — APPROVED** [review]

**Pull request merged**